torch/_inductor/codegen/cpp.py

```diff
@@ -1669,38 +1725,63 @@ def get_reduction_code_buffer(loops, is_suffix=True):
+                            )
+                            return suffix
                         else:
-                            return kernel.reduction_prefix
-                return None
+                            prefix = (
+                                (
+                                    kernel.reduction_prefix
+                                    + kernel.parallel_reduction_prefix,
+                                    kernel.local_reduction_init,
+                                    kernel.local_reduction_stores,
+                                )
+                                if loop.parallel
+                                else (kernel.reduction_prefix, None, None)
+                            )
+                            return prefix
 
             def gen_loops(loops: List[LoopLevel], in_reduction=False):
                 with contextlib.ExitStack() as stack_outer:
                     if loops:
                         loop = loops[0]
-                        if loop.is_reduction() and not in_reduction:
-                            reduction_prefix = get_reduction_code_buffer(
-                                loops, is_suffix=False
-                            )
+                        if loop.is_reduction and not in_reduction:
+                            (
+                                reduction_prefix,
+                                ws_init,
+                                ws_tores,
+                            ) = get_reduction_code_buffer(loops, is_suffix=False)
+                            if ws_init:
+                                assert ws_tores
+                                worksharing.reduction_init = ws_init
+                                worksharing.reduction_stores = ws_tores
                             if reduction_prefix:
                                 stack_outer.enter_context(code.indent())
                             code.splice(reduction_prefix)
                         if loop_nest.is_reduction_only() and loop.parallel:
                             worksharing.parallel(threads)
 
                     for loop in loops:
-                        gen_loop(loop, in_reduction)
+                        gen_loop(loop, loops[0].is_reduction and not in_reduction)
 
                     if loops:
                         loop = loops[0]
                         if loop_nest.is_reduction_only() and loop.parallel:
                             worksharing.close()
-                        if loop.is_reduction() and not in_reduction:
+                        if loop.is_reduction and not in_reduction:
                             code.splice(
-                                get_reduction_code_buffer(loops, is_suffix=True)
+                                get_reduction_code_buffer(loops, is_suffix=True)[0]
                             )
 
-            def gen_loop(loop: LoopLevel, in_reduction=False):
+            def gen_loop(loop: LoopLevel, update_reduction_store=False):
```

Comment:
why do you change the arg name?

Reply:
The `in_reduction` is not used in this function. So I removed this unused arg.
https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/cpp.py#L1700C43-L1713
I also need an arg to represent whether we need to update the `reduction_store` here. So I created a new arg named `update_reduction_store`.